File: python/linked/linked_lists_2.py
# see Remove Duplicates for List and Node type definitions


# O(n)
def kth_to_last_element(linked_list, k):
    # maintain a leading and lagging pointer
    # lag will trail the lead by k
    lead_node = linked_list.head
    lag_node = linked_list.head
    lead_index = 0
    lag_index = 0

    while lead_node:
        lead_node = lead_node.next
        lead_index += 1
        if lead_index > k:
            lag_node = lag_node.next
            lag_index += 1

    if lead_index - lag_index < k:
        print('List is not long enough to have a %d(st/nd/rd/th) to last element, returning the first node instead.')

    return lag_node

File: ts/solutions/linked/linked_lists_2.ts
import { LinkedList, Node_ } from './linked_lists_1'

// see Remove Duplicates for List and Node type definitions

// O(n)
function kthToLastElement(l: LinkedList, k: number): Node_ {
  let leadNode: Node_ = l.head
  let lagNode: Node_ = l.head
  let leadIndex = 0
  let lagIndex = 0

  while (leadNode) {
    leadNode = leadNode.next
    leadIndex++
    if (leadIndex > k) {
      lagNode = lagNode.next
      lagIndex++
    }
  }

  if (leadIndex - lagIndex < k) {
    console.log(
      `List is not long enough to have a ${k}(st/nd/rd/th) to last element, returning the first node instead.`,
    )
  }

  return lagNode
}
